File: ogl/src/main/java/com/github/ykiselev/opengl/matrices/Vector3f.java
package com.github.ykiselev.opengl.matrices;

/**
 * @author Yuriy Kiselev (uze@yandex.ru).
 */
public final class Vector3f {

    public float x, y, z;

    private enum Ops {

        ADDITION {
            @Override
            float apply(float a, float b) {
                return a + b;
            }
        },
        SUBTRACTION {
            @Override
            float apply(float a, float b) {
                return a - b;
            }
        },
        MULTIPLICATION {
            @Override
            float apply(float a, float b) {
                return a * b;
            }
        },
        DIVISION {
            @Override
            float apply(float a, float b) {
                return a / b;
            }
        };

        abstract float apply(float a, float b);
    }

    public Vector3f() {
    }

    public Vector3f(Vector3f src) {
        this(src.x, src.y, src.z);
    }

    public Vector3f(float x, float y, float z) {
        this.x = x;
        this.y = y;
        this.z = z;
    }

    public void set(float x, float y, float z) {
        this.x = x;
        this.y = y;
        this.z = z;
    }

    @Override
    public String toString() {
        return "Vector3f{" +
                "x=" + x +
                ", y=" + y +
                ", z=" + z +
                '}';
    }

    /**
     * Normalizes vector.
     */
    public void normalize() {
        final double length = length();
        double ool = 1.0;
        if (length != 0) {
            ool = 1.0 / length;
        }
        x *= ool;
        y *= ool;
        z *= ool;
    }

    /**
     * Calculates squared vector length.
     *
     * @return squared length of vector
     */
    public double squareLength() {
        return x * x + y * y + z * z;
    }

    /**
     * Calculates vector length.
     *
     * @return vector length
     */
    public double length() {
        return Math.sqrt(squareLength());
    }

    /**
     * Calculates Dot product between this vectors.
     *
     * @param b the second vector
     */
    public double dotProduct(Vector3f b) {
        return x * b.x + y * b.y + z * b.z;
    }

    /**
     * Calculates cross-product vector c = (a × b) that is perpendicular to both a and b, with a direction given by the right-hand rule.
     *
     * @param a the first vector
     * @param b the second vector
     */
    public void crossProduct(Vector3f a, Vector3f b) {
        set(
                a.y * b.z - a.z * b.y,
                a.z * b.x - a.x * b.z,
                a.x * b.y - a.y * b.x
        );
    }

    /**
     * Calculates cross-product of this vector and another.
     *
     * @param b the second vector
     */
    public void crossProduct(Vector3f b) {
        crossProduct(this, b);
    }

    /**
     * Scales the vector.
     */
    public void scale(float scale) {
        x *= scale;
        y *= scale;
        z *= scale;
    }

    /**
     * Applies specified operation to supplied vectors.
     *
     * @param a the first vector
     * @param b the second vector
     */
    private void apply(Vector3f a, Vector3f b, Ops op) {
        x = op.apply(a.x, b.x);
        y = op.apply(a.y, b.y);
        z = op.apply(a.z, b.z);
    }

    /**
     * Adds one vector to another.
     *
     * @param a the first vector
     * @param b the second vector
     */
    public void add(Vector3f a, Vector3f b) {
        apply(a, b, Ops.ADDITION);
    }

    public void add(Vector3f b) {
        add(this, b);
    }

    /**
     * Subtracts one vector from another.
     *
     * @param a the first vector
     * @param b the second vector
     */
    public void subtract(Vector3f a, Vector3f b) {
        apply(a, b, Ops.SUBTRACTION);
    }

    public void subtract(Vector3f b) {
        subtract(this, b);
    }

    /**
     * Multiplies one vector by another.
     *
     * @param a the first vector
     * @param b the second vector
     */
    public void multiply(Vector3f a, Vector3f b) {
        apply(a, b, Ops.MULTIPLICATION);
    }

    public void multiply(Vector3f b) {
        multiply(this, b);
    }

    /**
     * Divides one vector by another.
     *
     * @param a the first vector
     * @param b the second vector
     */
    public void divide(Vector3f a, Vector3f b) {
        apply(a, b, Ops.DIVISION);
    }

    public void divide(Vector3f b) {
        divide(this, b);
    }

    /**
     * Checks if two vectors are equal (absolute difference in each components is less than delta).
     *
     * @param b     the second vector
     * @param delta the maximum difference between two numbers for which they are still considered equal.
     * @return the true if vectors are equal or false otherwise.
     */
    public boolean equals(Vector3f b, float delta) {
        return equals(x, b.x, delta)
                && equals(y, b.y, delta)
                && equals(z, b.z, delta);
    }

    public boolean isEmpty(float delta) {
        return equals(x, 0, delta)
                && equals(y, 0, delta)
                && equals(z, 0, delta);
    }

    private boolean equals(float a, float b, float delta) {
        return Math.abs(a - b) <= delta + delta * Math.abs(b);
    }
}

File: assets/ogl/src/main/java/com/github/ykiselev/opengl/matrices/Vector3f.kt
/*
 * Copyright 2017 Yuriy Kiselev (uze@yandex.ru)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.github.ykiselev.opengl.matrices

import kotlin.math.abs
import kotlin.math.sqrt

/**
 * @author Yuriy Kiselev (uze@yandex.ru).
 */
data class Vector3f(
    @JvmField
    var x: Float = 0f,
    @JvmField
    var y: Float = 0f,
    @JvmField
    var z: Float = 0f
) {

    fun set(x: Float, y: Float, z: Float): Vector3f {
        this.x = x
        this.y = y
        this.z = z
        return this
    }

    fun set(src: Vector3f): Vector3f {
        this.x = src.x
        this.y = src.y
        this.z = src.z
        return this
    }

    /**
     * Calculates squared vector length.
     *
     * @return squared length of vector
     */
    fun squareLength(): Double {
        return (x * x + y * y + z * z).toDouble()
    }

    /**
     * Calculates vector length.
     *
     * @return vector length
     */
    fun length(): Double {
        return sqrt(squareLength())
    }

    /**
     * Calculates Dot product between this and supplied vector.
     *
     * @param b the second vector
     */
    infix fun dot(b: Vector3f): Double =
        x.toDouble() * b.x + y * b.y + z * b.z

    private enum class Ops {
        ADDITION {
            override fun apply(a: Float, b: Float): Float {
                return a + b
            }
        },
        SUBTRACTION {
            override fun apply(a: Float, b: Float): Float {
                return a - b
            }
        },
        MULTIPLICATION {
            override fun apply(a: Float, b: Float): Float {
                return a * b
            }
        },
        DIVISION {
            override fun apply(a: Float, b: Float): Float {
                return a / b
            }
        };

        abstract fun apply(a: Float, b: Float): Float
    }

    /**
     * Calculates cross-product vector c = (a × b) that is perpendicular to both a and b, with a direction given by the right-hand rule.
     *
     * @param b the second vector
     */
    infix fun cross(b: Vector3f): Vector3f =
        Vector3f(
            y * b.z - z * b.y,
            z * b.x - x * b.z,
            x * b.y - y * b.x
        )

    /**
     * Scales the vector.
     */
    operator fun times(scale: Float): Vector3f =
        Vector3f(x * scale, y * scale, z * scale)

    operator fun times(scale: Double): Vector3f =
        times(scale.toFloat())

    /**
     * Applies specified operation to this and supplied vectors.
     *
     * @param b the second vector
     */
    private fun apply(b: Vector3f, op: Ops): Vector3f =
        Vector3f(
            op.apply(x, b.x),
            op.apply(y, b.y),
            op.apply(z, b.z)
        )

    /**
     * Adds one vector to another.
     *
     * @param b the second vector
     */
    operator fun plus(b: Vector3f) =
        apply(b, Ops.ADDITION)

    /**
     * Subtracts one vector from another.
     *
     * @param b the second vector
     */
    operator fun minus(b: Vector3f): Vector3f =
        apply(b, Ops.SUBTRACTION)


    /**
     * Multiplies one vector by another.
     *
     * @param b the second vector
     */
    operator fun times(b: Vector3f) =
        apply(b, Ops.MULTIPLICATION)

    /**
     * Divides one vector by another.
     *
     * @param b the second vector
     */
    operator fun div(b: Vector3f) =
        apply(b, Ops.DIVISION)

    fun normalize(): Vector3f {
        val length = length()
        if (length != 0.0) {
            return this * (1.0 / length)
        }
        return copy()
    }

    /**
     * Checks if two vectors are equal (absolute difference in each component is less than delta).
     *
     * @param b     the second vector
     * @param delta the maximum difference between two numbers for which they are still considered equal.
     * @return the true if vectors are equal or false otherwise.
     */
    fun equals(b: Vector3f, delta: Float): Boolean =
        (equals(x, b.x, delta)
                && equals(y, b.y, delta)
                && equals(z, b.z, delta))

    fun isEmpty(delta: Float): Boolean =
        (equals(x, 0f, delta)
                && equals(y, 0f, delta)
                && equals(z, 0f, delta))

    private fun equals(a: Float, b: Float, delta: Float): Boolean =
        abs((a - b).toDouble()) <= delta + delta * abs(b.toDouble())
}
